File: graph-editor/src/lib/dslExplosion.ts
/**
 * DSL Explosion Utilities
 * 
 * Proper recursive descent parsing for compound query expressions.
 * 
 * Equivalences (all produce same atomic slices):
 * - (a;b).c = c.(a;b) = or(a,b).c = or(a.c,b.c) = a.c;b.c
 * - a;b;c = or(a,b,c)
 * - or(a,or(b,c)) = a;b;c
 */

import { parseConstraints, normalizeConstraintString } from './queryDSL';
import { contextRegistry } from '../services/contextRegistry';

/**
 * Explode compound DSL into atomic slices.
 */
/**
 * Explode compound DSL into atomic slices.
 * 
 * Uses parseConstraints from queryDSL.ts for all atomic expressions.
 * This ensures one standard parser throughout the codebase.
 * 
 * @param dsl - Compound expression (may have ;, or(), parentheses)
 * @returns Array of atomic slice DSL strings (normalized)
 */
export async function explodeDSL(dsl: string): Promise<string[]> {
  if (!dsl || !dsl.trim()) return [];
  
  // Step 1: Parse into branches (handles ;, or(), distribution)
  const branches = parseExpression(dsl);
  
  // Step 2: Expand bare keys in each branch (Cartesian product)
  const expanded: string[] = [];
  for (const branch of branches) {
    const slices = await expandBareKeys(branch);
    expanded.push(...slices);
  }
  
  // Step 3: Normalize all slices (uses queryDSL.normalizeConstraintString)
  return expanded.map(s => normalizeConstraintString(s));
}

/**
 * Parse DSL string into expression tree.
 */
function parseExpression(dsl: string): string[] {
  const trimmed = dsl.trim();
  
  // Handle or(...).suffix
  const startsWithOr = (() => {
    if (!trimmed.startsWith('or')) return false;
    let j = 2;
    while (j < trimmed.length && /\s/.test(trimmed[j])) j++;
    return trimmed[j] === '(';
  })();
  if (startsWithOr) {
    let openIndex = 2;
    while (openIndex < trimmed.length && /\s/.test(trimmed[openIndex])) openIndex++;
    const parenEnd = findMatchingParen(trimmed, openIndex);
    if (parenEnd === -1) {
      throw new Error(`Unbalanced parentheses in DSL near: ${trimmed}`);
    }
    const orPart = trimmed.substring(0, parenEnd + 1);
    const suffix = trimmed.substring(parenEnd + 1);
    
    const contents = extractFunctionContents(orPart, 'or');
    const parts = smartSplit(contents, ',');
    const branches: string[] = [];
    for (const part of parts) {
      const partBranches = parseExpression(part);
      // Apply suffix to each branch
      for (const branch of partBranches) {
        // Re-parse after suffix application so suffix expressions like `.or(...)` are handled.
        branches.push(...parseExpression(branch + suffix));
      }
    }
    return branches;
  }
  
  // Handle (...) - Check if outer parens match and can be stripped
  if (trimmed.startsWith('(') && !trimmed.startsWith('or(')) {
    const parenEnd = findMatchingParen(trimmed, 0);
    // If the matching paren is at the end, strip outer parens
    if (parenEnd === trimmed.length - 1) {
      const inner = trimmed.slice(1, -1);
      return parseExpression(inner);
    }
  }
  
  // Handle (...).suffix
  if (trimmed.startsWith('(') || trimmed.startsWith('or(')) {
    const parenEnd = findMatchingParen(trimmed, trimmed.indexOf('('));
    if (parenEnd < trimmed.length - 1 && trimmed[parenEnd + 1] === '.') {
      const prefix = trimmed.substring(0, parenEnd + 1);
      const suffix = trimmed.substring(parenEnd + 1);
      
      // Parse prefix, distribute suffix
      const prefixBranches = parseExpression(prefix);
      return prefixBranches.map(b => b + suffix);
    }
  }
  
  // Handle prefix.(...) 
  const dotParenIndex = (() => {
    let depth = 0;
    for (let i = 0; i < trimmed.length - 1; i++) {
      const ch = trimmed[i];
      if (ch === '(') depth++;
      else if (ch === ')') depth--;
      if (depth === 0 && trimmed.startsWith('.(', i)) return i;
    }
    return -1;
  })();
  if (dotParenIndex > 0) {
    const prefix = trimmed.substring(0, dotParenIndex);
    const rest = trimmed.substring(dotParenIndex + 1);
    
    const restBranches = parseExpression(rest);
    return restBranches.map(b => prefix + '.' + b);
  }

  // Handle prefix.or(...)
  // This is a common pattern in pinned DSLs: `or(cohort(...),window(...)).or(context(channel),context(geo))`
  // Semantics: distribute the OR branches as if written `prefix.(a;b)` / `or(prefix.a,prefix.b)`.
  //
  // Without this, the `.or(...)` remains as text and bare-key expansion can incorrectly
  // interpret `context(channel)` and `context(geo)` as simultaneous constraints (Cartesian product).
  const dotOrIndex = (() => {
    let depth = 0;
    for (let i = 0; i < trimmed.length - 4; i++) {
      const ch = trimmed[i];
      if (ch === '(') depth++;
      else if (ch === ')') depth--;
      if (depth !== 0) continue;
      if (trimmed[i] !== '.') continue;
      // Allow optional whitespace between '.or' and '('
      const rest = trimmed.substring(i + 1);
      if (!rest.startsWith('or')) continue;
      let j = i + 1 + 2; // after 'or'
      while (j < trimmed.length && /\s/.test(trimmed[j])) j++;
      if (trimmed[j] === '(') return i;
    }
    return -1;
  })();
  if (dotOrIndex > 0) {
    const prefix = trimmed.substring(0, dotOrIndex);
    const orStart = dotOrIndex + 1; // points at 'o' in 'or('
    const openParenIndex = trimmed.indexOf('(', orStart);
    const parenEnd = findMatchingParen(trimmed, openParenIndex);
    if (parenEnd === -1) {
      throw new Error(`Unbalanced parentheses in DSL near: ${trimmed.substring(orStart)}`);
    }
    const orPart = trimmed.substring(orStart, parenEnd + 1); // 'or(...)'
    const suffix = trimmed.substring(parenEnd + 1); // may be empty, or include further constraints

    const prefixBranches = parseExpression(prefix);
    const orBranches = parseExpression(orPart);

    const branches: string[] = [];
    for (const pb of prefixBranches) {
      for (const ob of orBranches) {
        branches.push(...parseExpression(`${pb}.${ob}${suffix}`));
      }
    }
    return branches;
  }
  
  // Handle semicolons at top level
  if (trimmed.includes(';')) {
    const parts = smartSplit(trimmed, ';');
    const branches: string[] = [];
    for (const part of parts) {
      branches.push(...parseExpression(part));
    }
    return branches;
  }
  
  // Atomic expression
  return [trimmed];
}

/**
 * Extract contents of function(...)
 */
function extractFunctionContents(str: string, funcName: string): string {
  const start = str.indexOf('(');
  if (start === -1) return '';
  
  const end = findMatchingParen(str, start);
  if (end === -1) {
    throw new Error(`Unbalanced parentheses in ${funcName}() expression: ${str}`);
  }
  return str.substring(start + 1, end);
}

/**
 * Find matching closing paren.
 */
function findMatchingParen(str: string, openIndex: number): number {
  let depth = 1;
  for (let i = openIndex + 1; i < str.length; i++) {
    if (str[i] === '(') depth++;
    else if (str[i] === ')') {
      depth--;
      if (depth === 0) return i;
    }
  }
  return -1;
}

/**
 * Split respecting parentheses.
 */
function smartSplit(str: string, sep: string): string[] {
  const parts: string[] = [];
  let current = '';
  let depth = 0;
  
  for (const char of str) {
    if (char === '(') depth++;
    else if (char === ')') depth--;
    else if (char === sep && depth === 0) {
      if (current.trim()) parts.push(current.trim());
      current = '';
      continue;
    }
    current += char;
  }
  
  if (current.trim()) parts.push(current.trim());
  return parts;
}

/**
 * Expand bare context keys (Cartesian product).
 * Uses parseConstraints from queryDSL.ts to parse each atomic expression.
 */
async function expandBareKeys(clause: string): Promise<string[]> {
  const parsed = parseConstraints(clause);
  const bareKeys = parsed.context.filter(ctx => !ctx.value);
  
  if (bareKeys.length === 0) {
    return [clause];
  }
  
  // Load all values for bare keys
  const keyValues: Array<{ key: string; values: any[] }> = [];
  for (const bareKey of bareKeys) {
    const values = await contextRegistry.getValuesForContext(bareKey.key);
    keyValues.push({ key: bareKey.key, values });
  }
  
  // Cartesian product
  const combinations = cartesianProduct(keyValues);
  
  const expanded: string[] = [];
  for (const combo of combinations) {
    let expandedClause = clause;
    for (const { key, value } of combo) {
      expandedClause = expandedClause.replace(
        new RegExp(`context\\(${key}\\)`, 'g'),
        `context(${key}:${value.id})`
      );
    }
    expanded.push(expandedClause);
  }
  
  return expanded;
}

/**
 * Cartesian product.
 */
function cartesianProduct(keyValues: Array<{ key: string; values: any[] }>): Array<Array<{ key: string; value: any }>> {
  if (keyValues.length === 0) return [[]];
  
  const [first, ...rest] = keyValues;
  const restProduct = cartesianProduct(rest);
  
  const result: Array<Array<{ key: string; value: any }>> = [];
  for (const value of first.values) {
    for (const combo of restProduct) {
      result.push([{ key: first.key, value }, ...combo]);
    }
  }
  
  return result;
}

/**
 * Count atomic slices.
 */
export async function countAtomicSlices(dsl: string): Promise<number> {
  const slices = await explodeDSL(dsl);
  return slices.length;
}

File: bayes/dsl_explosion.py
"""
DSL Explosion — Python port of graph-editor/src/lib/dslExplosion.ts.

Recursive descent parser for compound query DSL expressions.
Produces a list of atomic clause strings from compound expressions.

Equivalences (all produce the same atomic clauses):
    (a;b).c  =  c.(a;b)  =  or(a,b).c  =  or(a.c,b.c)  =  a.c;b.c
    a;b;c    =  or(a,b,c)
    or(a,or(b,c))  =  a;b;c

This module handles structural explosion only (semicolons, or(),
parenthesised groups, dot-distribution).  Bare-key expansion
(context(channel) → one clause per value) is NOT included because it
requires runtime access to the context registry; callers that need it
must expand bare keys separately after calling explode_dsl().
"""

from __future__ import annotations


def explode_dsl(dsl: str) -> list[str]:
    """Explode a compound DSL string into atomic clause strings.

    Returns a list of non-empty, stripped clause strings.
    Empty input returns [].
    """
    if not dsl or not dsl.strip():
        return []
    branches = _parse_expression(dsl.strip())
    # Filter empty branches, strip whitespace
    return [b.strip() for b in branches if b.strip()]


# ---------------------------------------------------------------------------
# Core recursive descent parser
# ---------------------------------------------------------------------------

def _parse_expression(dsl: str) -> list[str]:
    trimmed = dsl.strip()
    if not trimmed:
        return [trimmed]

    # --- or(...) possibly followed by suffix ---
    if _starts_with_or(trimmed):
        open_idx = trimmed.index("(", 2)
        paren_end = _find_matching_paren(trimmed, open_idx)
        if paren_end == -1:
            raise ValueError(f"Unbalanced parentheses in DSL near: {trimmed}")
        or_part = trimmed[:paren_end + 1]
        suffix = trimmed[paren_end + 1:]

        contents = _extract_function_contents(or_part, "or")
        parts = _smart_split(contents, ",", keep_empty=True)
        branches: list[str] = []
        for part in parts:
            if part == "":
                # Empty part in or() → uncontexted branch
                branches.append(suffix[1:] if suffix.startswith(".") else (suffix or ""))
            else:
                part_branches = _parse_expression(part)
                for branch in part_branches:
                    branches.extend(_parse_expression(branch + suffix))
        return branches

    # --- (...) outer parens that wrap the entire string → strip ---
    if trimmed.startswith("(") and not trimmed.startswith("or("):
        paren_end = _find_matching_paren(trimmed, 0)
        if paren_end == len(trimmed) - 1:
            inner = trimmed[1:-1]
            return _parse_expression(inner)

    # --- (...)suffix or (...)(suffix) → cartesian distribution ---
    if trimmed.startswith("(") or trimmed.startswith("or("):
        open_idx = trimmed.index("(")
        paren_end = _find_matching_paren(trimmed, open_idx)
        if paren_end != -1 and paren_end < len(trimmed) - 1:
            next_char = trimmed[paren_end + 1]
            if next_char in (".", "("):
                prefix = trimmed[:paren_end + 1]
                # For (...)(suffix), insert a dot so distribution works uniformly
                if next_char == "(":
                    suffix = "." + trimmed[paren_end + 1:]
                else:
                    suffix = trimmed[paren_end + 1:]

                prefix_branches = _parse_expression(prefix)
                branches = []
                for b in prefix_branches:
                    branches.extend(_parse_expression(b + suffix))
                return branches

    # --- prefix.(...) ---
    dot_paren_idx = _find_dot_paren(trimmed)
    if dot_paren_idx > 0:
        prefix = trimmed[:dot_paren_idx]
        rest = trimmed[dot_paren_idx + 1:]  # skip the dot
        rest_branches = _parse_expression(rest)
        return [prefix if b == "" else f"{prefix}.{b}" for b in rest_branches]

    # --- prefix.or(...) ---
    dot_or_idx = _find_dot_or(trimmed)
    if dot_or_idx > 0:
        prefix = trimmed[:dot_or_idx]
        or_start = dot_or_idx + 1  # points at 'o' in 'or('
        open_paren_idx = trimmed.index("(", or_start)
        paren_end = _find_matching_paren(trimmed, open_paren_idx)
        if paren_end == -1:
            raise ValueError(
                f"Unbalanced parentheses in DSL near: {trimmed[or_start:]}"
            )
        or_part = trimmed[or_start:paren_end + 1]
        suffix = trimmed[paren_end + 1:]

        prefix_branches = _parse_expression(prefix)
        or_branches = _parse_expression(or_part)

        branches = []
        for pb in prefix_branches:
            for ob in or_branches:
                branches.extend(_parse_expression(f"{pb}.{ob}{suffix}"))
        return branches

    # --- Top-level semicolons ---
    if ";" in trimmed:
        parts = _smart_split(trimmed, ";", keep_empty=True)
        branches = []
        for part in parts:
            if part == "":
                branches.append("")
            else:
                branches.extend(_parse_expression(part))
        return branches

    # --- Atomic expression ---
    return [trimmed]


# ---------------------------------------------------------------------------
# Helper: detect or( at start, allowing whitespace between 'or' and '('
# ---------------------------------------------------------------------------

def _starts_with_or(s: str) -> bool:
    if not s.startswith("or"):
        return False
    j = 2
    while j < len(s) and s[j].isspace():
        j += 1
    return j < len(s) and s[j] == "("


# ---------------------------------------------------------------------------
# Helper: find prefix.( pattern at depth 0
# ---------------------------------------------------------------------------

def _find_dot_paren(s: str) -> int:
    depth = 0
    for i in range(len(s) - 1):
        ch = s[i]
        if ch == "(":
            depth += 1
        elif ch == ")":
            depth -= 1
        if depth == 0 and s[i:i + 2] == ".(":
            # Make sure this isn't .or(
            rest = s[i + 1:]
            if rest.startswith("or"):
                j = 2
                while j < len(rest) and rest[j].isspace():
                    j += 1
                if j < len(rest) and rest[j] == "(":
                    continue  # This is .or(...), not .(...)
            return i
    return -1


# ---------------------------------------------------------------------------
# Helper: find prefix.or( pattern at depth 0
# ---------------------------------------------------------------------------

def _find_dot_or(s: str) -> int:
    depth = 0
    for i in range(len(s) - 4):
        ch = s[i]
        if ch == "(":
            depth += 1
        elif ch == ")":
            depth -= 1
        if depth != 0:
            continue
        if s[i] != ".":
            continue
        rest = s[i + 1:]
        if not rest.startswith("or"):
            continue
        j = 2
        while j < len(rest) and rest[j].isspace():
            j += 1
        if j < len(rest) and rest[j] == "(":
            return i
    return -1


# ---------------------------------------------------------------------------
# Parenthesis matching
# ---------------------------------------------------------------------------

def _find_matching_paren(s: str, open_index: int) -> int:
    depth = 1
    for i in range(open_index + 1, len(s)):
        if s[i] == "(":
            depth += 1
        elif s[i] == ")":
            depth -= 1
            if depth == 0:
                return i
    return -1


# ---------------------------------------------------------------------------
# Extract contents of funcname(...)
# ---------------------------------------------------------------------------

def _extract_function_contents(s: str, func_name: str) -> str:
    start = s.index("(")
    end = _find_matching_paren(s, start)
    if end == -1:
        raise ValueError(
            f"Unbalanced parentheses in {func_name}() expression: {s}"
        )
    return s[start + 1:end]


# ---------------------------------------------------------------------------
# Paren-aware split on a single-char separator
# ---------------------------------------------------------------------------

def _smart_split(s: str, sep: str, *, keep_empty: bool = False) -> list[str]:
    parts: list[str] = []
    current: list[str] = []
    depth = 0

    for ch in s:
        if ch == "(":
            depth += 1
            current.append(ch)
        elif ch == ")":
            depth -= 1
            current.append(ch)
        elif ch == sep and depth == 0:
            piece = "".join(current).strip()
            if keep_empty or piece:
                parts.append(piece)
            current = []
            continue
        else:
            current.append(ch)

    piece = "".join(current).strip()
    if keep_empty or piece:
        parts.append(piece)
    return parts
